File: LeetCode/Python/0404. Sum of Left Leaves.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def is_leaf(self, root):
        return root and not root.left and not root.right
        
    def sumOfLeftLeaves(self, root: Optional[TreeNode], is_left=False) -> int:
        if not root:
            return 0
        if self.is_leaf(root) and is_left:
            return root.val
        return self.sumOfLeftLeaves(root.left, True) + self.sumOfLeftLeaves(root.right, False)

File: LeetCode/Java/0404. Sum of Left Leaves.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 
class Solution {
    private boolean isLeaf(TreeNode root) {
        return root != null && root.left == null && root.right == null;
    }
    
    private int sumOfLeftLeaves(TreeNode root, boolean isLeft) {
        if (root == null)
            return 0;
        if (isLeaf(root) && isLeft)
            return root.val;
        return sumOfLeftLeaves(root.left, true) + sumOfLeftLeaves(root.right, false);
    }
    
    public int sumOfLeftLeaves(TreeNode root) {
        return sumOfLeftLeaves(root, false);
    } 
}
